vta/src/tsim/virtual_memory.cc

```diff
@@ -0,0 +1,185 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "virtual_memory.h"
+
+#include <dmlc/logging.h>
+#include <cstdint>
+#include <cstdlib>
+#include <cstring>
+#include <list>
+#include <utility>
+#include <iterator>
+#include <unordered_map>
+
+/*! page size of virtual address */
+#ifndef VTA_TSIM_VM_PAGE_SIZE
+#define VTA_TSIM_VM_PAGE_SIZE (4096)
+#endif  // VTA_TSIM_VM_PAGE_SIZE
+
+/*! starting point of the virtual address system */
+#ifndef VTA_TSIM_VM_ADDR_BEGIN
+#define VTA_TSIM_VM_ADDR_BEGIN (0x40000000)
+#endif  // VTA_TSIM_VM_ADDR_BEGIN
+
+namespace vta {
+namespace tsim {
```

Comment:
See also https://github.com/dmlc/tvm/blob/master/vta/src/sim/sim_driver.cc#L129. perhaps we can reuse code between the two

Reply:
I have already exposed the interface for memory allocation, release and memcpy in the header file, which doesn't contain namespaces. Those functions could be potentially reused in the `sim_driver`, however, I think it's more important for TSIM to test with PYNQ (see #3380).